cpp/src/arrow/result.h

```diff
@@ -212,9 +240,15 @@ class Result : public util::EqualityComparable<Result<T>> {
   template <typename U,
             typename E = typename std::enable_if<std::is_constructible<T, U&&>::value &&
                                                  std::is_convertible<U, T>::value>::type>
-  Result(Result<U>&& other) : variant_("unitialized") {
-    AssignVariant(std::move(other.variant_));
-    other.variant_ = "Value was moved to another Result.";
+  Result(Result<U>&& other) noexcept {
+    if (ARROW_PREDICT_TRUE(other.status_.ok())) {
```

Comment:
Can you just:
```c++
std::swap(status_, other.status_);
std::swap(data_, other.data_);
```
That would remove the destructor crash issue.

Reply:
`std::swap(data_, other.data_);` will only work if `T` can be trivially moved using `memcpy`.

Reply:
And actually, it doesn't work at all, since `*this` isn't initialized at this point, so you're filling `other` with junk.